### ISSUE TITLE
Uploading big file from Mac sometimes get stuck at requestStream.Close()

### ISSUE THREAD

**Issue by nicolas-raoul** [OpenDataSpace/chemistry-dotcmis#2]
Hello,

When I upload a 400MB or 40MB file using OpenDataSpace's DotCMIS from Mac OS X, the upload often (50% probability) gets stuck at line 234 in `http.cs`:

```
                    // send data
                    if (writer != null)
                    {
                        conn.SendChunked = true;
                        Stream requestStream = conn.GetRequestStream();
                        writer(requestStream);
                        requestStream.Close();          <- Gets stuck here
                    }
```

In those cases, the program just gets stuck there and never ends.

I have created a small proof-of-concept that shows the bug:
https://github.com/nicolas-raoul/dotcmis-upload
Just replace the URL/folder/username/password, compile and execute.

Tested with Alfresco 4.2.e on Mac OS X 10.9 and Mac OS X 10.10

Thank you!
Nicolas

**Comment by soneyworld** [comment]
Hi Nicolas,

which version of mono do you use? There was/is a bug in mono which blocks a close operation until forever.

Best regards

**Comment by nicolas-raoul** [comment]
I use Mono 3.10.0
What version solves the problem?
Thanks a lot!

```
$ mono -V
Mono JIT compiler version 3.10.0 ((detached/e204655 Mon Oct 13 08:50:28 EDT 2014)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
    TLS:           normal
    SIGSEGV:       altstack
    Notification:  kqueue
    Architecture:  x86
    Disabled:      none
    Misc:          softdebug 
    LLVM:          yes(3.4svn-mono-(detached/e656cac)
    GC:            sgen
```

**Comment by soneyworld** [comment]
Thanks a lot! That is weird. It should be fixed since 3.6. I will verify it with your provided code!

**Comment by soneyworld** [comment]
Yes, you are right, I could reproduce it with your test case. I will file a bug to mono.
Thanks a lot!

**Comment by nicolas-raoul** [comment]
This bug does not affect you in DataSpace Sync?
Do you upload documents using another method?

**Comment by soneyworld** [comment]
It does and I also filed the bug https://bugzilla.xamarin.com/show_bug.cgi?id=21135 again. I hope they will fix it as fast as possible. For Linux, we built our own mono where this bug should be fixed:
https://build.opensuse.org/package/show/home:felfert/mono-core

**Comment by nicolas-raoul** [comment]
How about disabling chunking for upload when we detect that the plaftform is Mono?
Would it work better?

I am currently trying to disable chunking, but unfortunately there does not seem to be any easy way to disable it, I would have to modify OpenDataSpaceDotCMIS quite heavily... or is there a better way to do it?

**Comment by soneyworld** [comment]
That's the reason why I didn't disabled chunked upload. But if you find a way to disable it, I would merge it back. The best way would be a fixed mono, but it seems to be a low priority for Xamarin.

**Comment by nicolas-raoul** [comment]
I just wrote this patch, it solves the upload problem by disabling upload chunking:

```
diff --git a/DotCMIS/DotCMIS.csproj b/DotCMIS/DotCMIS.csproj
index 411a31b..d2e2d78 100644
--- a/DotCMIS/DotCMIS.csproj
+++ b/DotCMIS/DotCMIS.csproj
@@ -10,7 +10,6 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>^M
     <RootNamespace>DotCMIS</RootNamespace>^M
     <AssemblyName>DotCMIS</AssemblyName>^M
-    <TargetFrameworkVersion>v3.5</TargetFrameworkVersion>^M
     <FileAlignment>512</FileAlignment>^M
   </PropertyGroup>^M
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">^M
diff --git a/DotCMIS/binding/http.cs b/DotCMIS/binding/http.cs
index c92e3dd..3a0f029 100644
--- a/DotCMIS/binding/http.cs
+++ b/DotCMIS/binding/http.cs
@@ -179,9 +179,25 @@ namespace DotCMIS.Binding.Impl
                     // send data^M
                     if (writer != null)^M
                     {^M
+#if __MonoCS__ // Unchunked upload for Mono, because of Mono bug 21135^M
+^M
+                                               // First write request to temporary file, inefficient but easier than rewriting large parts of DotCMIS^M
+                                               string tempFile = System.IO.Path.GetTempFileName();^M
+                                               using (var tempStream = new StreamWriter(tempFile))^M
+                                               {^M
+                                                       writer(tempStream.BaseStream);^M
+                                               }^M
+^M
+                                               // Send the request to the server^M
+                                               FileStream tempFileStream = new FileStream(tempFile, FileMode.Open, FileAccess.Read);^M
+                                               conn.ContentLength = tempFileStream.Length;^M
+                                               Stream requestStream = conn.GetRequestStream();^M
+                                               tempFileStream.CopyTo(requestStream);^M
+#else^M
                                                conn.SendChunked = true;^M
                                                Stream requestStream = conn.GetRequestStream();^M
                                                writer(requestStream);^M
+#endif^M
                         requestStream.Close();^M
                     }^M
                     else^M
```

It is not optimized (actually very wasteful) but it works as a temporary fix while waiting for Mono to fix their bug.
Do you think you could include it in your DotCMIS? Thanks a lot!
`FileStream.CopyTo` needs .NET 4.0 but it is not complicated to implement for .NET 3.5 if you prefer.

**Comment by soneyworld** [comment]
Hi Nicolas,
Thanks a lot!

But there are a few problems in your patch:
1. There is no cleanup call to delete the temporary file after successful upload (very easy to fix).
2. And streamed large files are stored on the hard disk which could be a security issue (plain text content)
3. File uploads of really large files do need much more disk space (Uploading a 100 GB file from a 120GB HDD is impossible with this patch).
4. The upload progress is not detectable anymore by the library user.

So at the moment, I won't apply your patch, because we do need the correct progress.
And the points above are the reason why I haven't disabled the chunked upload. Last but not leased: It was possible to upload a file while it was written to the hard disk and not finished yet (writing and uploading in parallel).

Best regards

**Comment by nicolas-raoul** [comment]
Thank you for the feedback!
I fixed points 1 and 2 in my new patch below.
The other points could indeed be improved, but that is a temporary workaround while waiting for Mono to merge Fritz's pull request, which I hope will be soon :-)

I understand that this temporary workaround might not be worth merging into OpenDataSpace's DotCMIS, so I will create a friendly fork, hoping we can switch back to your implementation soon :-)

Cheers!
Nicolas

```
diff --git a/DotCMIS/DotCMIS.csproj b/DotCMIS/DotCMIS.csproj
index 411a31b..fac64d3 100644
--- a/DotCMIS/DotCMIS.csproj
+++ b/DotCMIS/DotCMIS.csproj
@@ -10,8 +10,9 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>^M
     <RootNamespace>DotCMIS</RootNamespace>^M
     <AssemblyName>DotCMIS</AssemblyName>^M
-    <TargetFrameworkVersion>v3.5</TargetFrameworkVersion>^M
     <FileAlignment>512</FileAlignment>^M
   </PropertyGroup>^M
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">^M
     <DebugSymbols>true</DebugSymbols>^M
diff --git a/DotCMIS/binding/http.cs b/DotCMIS/binding/http.cs
index c92e3dd..163f81b 100644
--- a/DotCMIS/binding/http.cs
+++ b/DotCMIS/binding/http.cs
@@ -179,10 +179,34 @@ namespace DotCMIS.Binding.Impl
                     // send data^M
                     if (writer != null)^M
                     {^M
+#if __MonoCS__ // Unchunked upload for Mono, because of Mono bug 21135^M
+^M
+                                               // First write request to temporary file, inefficient but easier than rewriting large parts of DotCMIS^M
+                                               string tempFile = System.IO.Path.GetTempFileName();^M
+                                               // Make the temporary file readable only by the user, since data might be confidential.^M
+                                               Mono.Unix.Native.Syscall.chmod(tempFile,^M
+                                                       Mono.Unix.Native.FilePermissions.S_IWUSR | Mono.Unix.Native.FilePermissions.S_IRUSR);^M
+                                               using (var tempStream = new StreamWriter(tempFile))^M
+                                               {^M
+                                                       writer(tempStream.BaseStream);^M
+                                               }^M
+^M
+                                               // Send the request to the server^M
+                                               FileStream tempFileStream = new FileStream(tempFile, FileMode.Open, FileAccess.Read);^M
+                                               conn.ContentLength = tempFileStream.Length;^M
+                                               Stream requestStream = conn.GetRequestStream();^M
+                                               tempFileStream.CopyTo(requestStream);^M
+                                               requestStream.Close();^M
+^M
+                                               // Remove temp file^M
+                                               tempFileStream.Close();^M
+                                               File.Delete(tempFile);^M
+#else^M
                                                conn.SendChunked = true;^M
                                                Stream requestStream = conn.GetRequestStream();^M
                                                writer(requestStream);^M
                                                requestStream.Close();^M
+#endif^M
                     }^M
                     else^M
                     {^M
```

**Comment by soneyworld** [comment]
We fixed this issue by a patched self compiled mono version and will do so until our patch will reach the next stable mono release.
In detail: We created our own System.dll and installed it on the MacOS development machine into the GAC. That includes the modified System.dll into the monolithic MacOS App package.
Nearly the same on Linux, where we use our own patched version (https://build.opensuse.org/project/show/home:ods) of the provided mono deb/rpm packages.

Best regards
